### ISSUE TITLE
Idea Collection

### ISSUE THREAD

**Issue by mousyball** [mousyball/pytorch-trainer#1]
## Intro.

The arch is similar to [pytorch-lighting](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/docs/source/_images/general/pl_overview_flat.jpg).

## Arch

* (#3)Trainer
  * Trainer Setup
  * Data Setup
  * Training Loop
    * Train
    * Eval
    * Callback Hook (logger, tensorboard, ...)

## Components

> ToyNet and ToyDataset are needed for development.

* Registry modules: [[ref] mmcv](https://github.com/mousyball/pytorch-trainer/issues/new)
  * Transform
  * Network
  * Loss
  * Callback

* Logger
  * Design a logger class used to be called at the hook.

* Config: [[ref] fvcore](https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/config.py)
  * Use `yacs` to construct the config with inheritance.

* Loss
  * Design a class that includes multiple meters for monitoring loss.
  * Support the loss depended on tasks in examples or common scenarios.
  * Support
    * OD sampler could be referenced from mmdetection.
    * OHEM for detection/segmentation/other tasks.

* LR Schedular
  * Plain
  * Stepwise
  * Cosine Decay Restart

* Optimizer
  * SGD, Adam, Ranger, ...

* Loader for model weights
  * Exception handling
  * Layer by layer loading

* Network
  * SyncBN
  * custom_freeze / freeze_bn / eval mode
  * lr_group and group weights
  * Builder?

* Tensorboard
  * loss
  * lr
  * inference result of an image
  * confused prediction

* Dataloader / Dataset
  * Support Public Dataset

## Features

* Resume checkpoints

* Integrate with `MLFlow`

* Utility functions
  * Cython helper function

* Distributed Training for GPU

* lr_finder

* Visualizer
  * Transform
  * Output (may depend on tasks)

* Evaluation for test set

* Demo Script

* Support ONNX

## Others

* Testing
  * `pytest`

* Environment
  * `.gitmessage`
  * `docker`
  * `pipenv`
    * requirements.txt

* Documentation
  * Readme
  * Describe the reason why the design is made from/for.
  * Dev SOP

**Comment by mousyball** [comment]
HI

**Comment by mousyball** [comment]
## These components should be considered first

* Registry module: [mmcv/mmcv/utils/registry.py](https://github.com/open-mmlab/mmcv/blob/master/mmcv/utils/registry.py)
  * [mmcv/docs/registry.md](https://github.com/open-mmlab/mmcv/blob/master/docs/registry.md): Check how it works.

* runner: [mmcv/mmcv/runner/](https://github.com/open-mmlab/mmcv/tree/master/mmcv/runner)
  * iteration-based
  * epoch-based

* hook: [mmcv/mmcv/runner/hooks/](https://github.com/open-mmlab/mmcv/tree/master/mmcv/runner/hooks)
  * Check the call order.

* logger: [mmcv/mmcv/runner/hooks/logger/](https://github.com/open-mmlab/mmcv/tree/master/mmcv/runner/hooks/logger)

* Directory Structure

**Comment by mousyball** [comment]
## NOTE

* Make sure that functions are de-coupled in the Callback.
* Runner has 2 modes: iterative and epoch.
* Schedular should re-write or use pytorch function?
* get_10x_lr_params checker
  * use module.paramters() to check
* batch_size finder
  * iteratively increase batch size and run training loop
* distributed train 
  * Sync Buffers between machine
